### ISSUE TITLE
App get stuck on linux when using MultipartFormDataContent

### ISSUE THREAD

**Issue by agoretsky** [aspnet/Mvc#5406]
I have server asp.net core 1.0.1 app with controller with method:

        [HttpPost]
        public IActionResult UploadImage()
        {
            this.logger.Debug("=====1");
            try
            {
                if (this.Request.Form.Files.Count != 1)
                {
                    this.logger.Debug("=====2");
                }
            }
            catch (Exception ex)
            {
                this.logger.Debug(ex.ToString);
            }

            this.logger.Debug("=====3");

            return this.Ok();
        }

And from client app I call this

        var bytes = File.ReadAllBytes(@"C:\Users\user\Pictures\test1.jpg");
        UploadImage(bytes);

        public static Image UploadImage(byte[] imageBytes)
        {

            HttpClient httpClient = new HttpClient();
            httpClient.DefaultRequestHeaders.Add("Bearer", Bearer);
            var content = new ByteArrayContent(imageBytes, 0, imageBytes.Length);
            MultipartFormDataContent form = new MultipartFormDataContent
            {
                {content, "picture", "hello1.jpg"}
            };
            content.Headers.ContentType = new MediaTypeHeaderValue("image/jpg");
            
            HttpResponseMessage response = httpClient.PostAsync(MediaUrl, form).Result;

            response.EnsureSuccessStatusCode();
        }

I can't see logs after checking this.Request.Form.Files.Count != 1.
Client app gets timeout.

I also using middleware before mvc, which doesn't produce logs too:

            app.Use(async (context, next) =>
            {
                await next.Invoke();
                this.logger.Trace("RESULT");
                this.logger.Trace(context.Response.StatusCode);
            });
App just get stuck on line if (this.Request.Form.Files.Count != 1)

This problem happens only on linux, when I run server app on local windows machine everything is ok.

**Comment by davidfowl** [comment]
Would it be possible for you to put a repro project on github?

**Comment by agoretsky** [comment]
@davidfowl, I figured out why it happened: on the linux server admin configured all requests (which starts with particular relative url prefix) to go through apache traffic server, he expected that these url will be used only for GET requests. So, when I did post request with file, ATS passed this request to asp.net core backend in incorrect state maybe, I don't know,  this results that when I tried access this.Request.Form.Files.Count property app just get in stuck, no errors, no throws, no logs. I don't know what app was doing. Now, without ATS, everything is ok. Thank you

**Comment by davidfowl** [comment]
@agoretsky would it be possible for you to get the raw request? It would be interesting to know what caused this (it sounds like this bug in Kestrel https://github.com/aspnet/KestrelHttpServer/commit/73656f6503688b6eb47ed7a5e93b3db76cbb2280)

**Comment by agoretsky** [comment]
@davidfowl, I will try to do it this week and then leave it here

**Comment by agoretsky** [comment]
@davidfowl,
I did it in controller:

           this.logger.Debug("=====0");

            using (var reader = new StreamReader(this.Request.Body))
            {
                var result = reader.ReadToEnd();
                this.logger.Trace(result);
            }

            this.logger.Debug("=====1");

Then I immediately received first log and then after 15 minutes using exception filter I've got it:

System.AggregateException: One or more errors occurred. (Unexpected end of request content) ---> Microsoft.AspNetCore.Server.Kestrel.BadHttpRequestException: Unexpected end of request content
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.MessageBody.ForContentLength.<ReadAsyncAwaited>d__4.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.FrameRequestStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.StreamReader.ReadBuffer()
   at System.IO.StreamReader.ReadToEnd()
---> (Inner Exception #0) Microsoft.AspNetCore.Server.Kestrel.BadHttpRequestException: Unexpected end of request content
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.MessageBody.ForContentLength.<ReadAsyncAwaited>d__4.MoveNext()<---

I did it twice and uploaded tcp dumps here: https://drive.google.com/open?id=0B_hcLdxD3U8TVTBrYVpsMUZaUU0. (ATS - 10.0.0.231, dotnet: 10.0.0.112, 10.0.0.111)

**Comment by halter73** [comment]
@agoretsky I took a look at stage-back1.pcab, and it looks like the client sent an incomplete request.

First I used `ip.src == <CLIENT IP>` as my display filter to narrow the packets down to the request in question. All of the request headers can be found in packet No. 40:

    POST /media/images HTTP/1.1
    Bearer: ########-####-####-####-############
    Content-Type: multipart/form-data; boundary="3be0e46e-6e5b-4b2b-9675-a537f4192589"
    Host: ####################
    Content-Length: 145311
    Expect: 100-continue
    X-Forwarded-For: #########, <CLIENT IP>
    Client-ip: <CLIENT IP>
    Via: http/1.1 ############### (ApacheTrafficServer/5.3.0)

In packet No. 42 Kestrel responds with `HTTP/1.1 100 Continue\r\n\r\n` as it should. From there the client sends all the request body it sends for the duration of the trace in packet Nos. 53, 54, 57, 62 and 63. All of these packets are sent with a tenth of a second starting 6.4 seconds into the 88 second trace.

The TCP segment data of those 5 packets are 10136, 4344, 1989, 15928 and 371 bytes respectively for a total of 32768 bytes. This only makes up for part of the 145311 bytes expected given the Content-Length header.

Over the 88 second trace, no packets from <CLIENT IP> include RST or FIN requests indicating that the request should be terminated, so Kestrel is waiting for the remaining 112543 it expects in the content body.

Eventually the OS likely kills the connection that the incomplete request was sent over causing Kestrel to fail with the `BadHttpRequestException: Unexpected end of request content`.
